java/src/main/native/src/TableJni.cpp

```diff
@@ -775,6 +777,31 @@ jlongArray combine_join_results(JNIEnv *env, cudf::table &left_results,
   return combine_join_results(env, std::move(left_cols), std::move(right_cols));
 }
 
+struct scalar_construction_helper {
+  template <typename T,
+           typename ScalarType = cudf::scalar_type_t<T>,
+           typename std::enable_if_t<!(std::is_integral<T>() && !cudf::is_boolean<T>())>* = nullptr>
+  std::unique_ptr<cudf::scalar> operator()(int value, bool validify = true) {
+    CUDF_FAIL("Range window only supports non-bool integral type");
+  }
+
+  template <typename T,
+            typename ScalarType = cudf::scalar_type_t<T>,
+            typename std::enable_if_t<std::is_integral<T>() && !cudf::is_boolean<T>()>* = nullptr>
+  std::unique_ptr<cudf::scalar> operator()(int value, bool validify = true) {
+    using Type = device_storage_type_t<T>;
+    auto s     = new ScalarType(Type{}, validify);
+    s->set_value(static_cast<T> (value));
+    return std::unique_ptr<scalar>(s);
+  }
+};
+
+std::unique_ptr<cudf::scalar> make_range_window_scalar(data_type type,
```

Comment:
I think we might not need this factory method at all. 

I'm not too familiar with the JNI scalar construction APIs any more. But I think think we should be able to use those to construct the bounds scalars.
For instance, if the window definition is like:
```sql
... RANGE BETWEEN 15 PRECEDING AND 5 FOLLOWING...
```
The `rapids-4-spark` plugin would already know the type corresponding to `15` and `5`. We should be able to use the JNI scalar APIs to construct these scalars and pass their addresses over JNI. (Sorry, I'll need to examine the code more closely to familiarize myself with JNI scalars.)

Reply:
Hi @mythrocks, you're right, I am doing that. Thx.

Reply:
Done